## Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants.
Justifica la teva elecció.  

->[Airlines Delay: Airline on-time statistics and delay causes](https://www.kaggle.com/giovamata/airlinedelaycauses)

**Important**:  
Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.  

Importem l'arxiu amb les dades de vols amb pandas:

In [1]:
import pandas as pd
df_flights = pd.read_csv(r'C:\Users\WotanJr\Desktop\Data Science - IT Academy\Sprint 3\estructures_Dataframe\DelayedFlights.csv', index_col=0)

df_flights.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   2003.0        1955   2211.0   
1  2008      1           3          4    754.0         735   1002.0   
2  2008      1           3          4    628.0         620    804.0   
4  2008      1           3          4   1829.0        1755   1959.0   
5  2008      1           3          4   1940.0        1915   2121.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   
4        1925            WN       3920  ...    3.0     10.0          0   
5        2110            WN        378  ...    4.0     10.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0           NaN          NaN      NaN   
1                 N         0           NaN          NaN      NaN   
2                 N         0           NaN          NaN      NaN   
4                 N         0           2.0          0.0      0.0   
5                 N         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
4            0.0               32.0  
5            NaN                NaN  

[5 rows x 29 columns]

In [2]:
df_flights.tail()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
7009710  2008     12          13          6   1250.0        1220   1617.0   
7009717  2008     12          13          6    657.0         600    904.0   
7009718  2008     12          13          6   1007.0         847   1149.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
7009710        1552            DL       1621  ...    9.0     18.0          0   
7009717         749            DL       1631  ...   15.0     34.0          0   
7009718        1010            DL       1631  ...    8.0     32.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
7009710                 N         0           3.0          0.0      0.0   
7009717                 N         0           0.0         57.0     18.0   
7009718                 N         0           1.0          0.0     19.0   
7009726                 N         0           NaN          NaN      NaN   
7009727                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
7009710            0.0               22.0  
7009717            0.0                0.0  
7009718            0.0               79.0  
7009726            NaN                NaN  
7009727            NaN                NaN  

[5 rows x 29 columns]

In [3]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1936758 entries, 0 to 7009727
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             float64
 20  TaxiOut            float64
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       float64
 25  WeatherDelay      

In [4]:
df_flights.shape

(1936758, 29)

### VARIABLES
En aquest projecte de kaggle hem trobat una definició exhaustiva de totes les columnes d'aquest dataset:

https://www.kaggle.com/code/adveros/flight-delay-eda-exploratory-data-analysis

Es fa notar que un vol es considera amb retard si el retard a l'arribada (columna "ArrDelay") és superior a 15 minuts.

Podem fer un resum de les 29 columnes:

Year: 2008  
Month: 1-12  
DayofMonth: 1-31  
DayOfWeek: 1 (Monday) - 7 (Sunday)  
DepTime: Hora de sortida real (local, hhmm)  
CRSDepTime: Hora de sortida programada (local, hhmm)  
ArrTime: Hora d'arribada real (local, hhmm)  
CRSArrTime: Hora d'arribada programada (local, hhmm)  
UniqueCarrier: Codi d'operador únic  
FlightNum: Número de vol  
TailNum: Matrícula de l'avió  
ActualElapsedTime: Temps transcorregut real (min)  
CRSElapsedTime: Temps transcorregut programat (min)  
AirTime: Temps en l'aire (min)  
ArrDelay: Retard en l'arribada (min)  
DepDelay: Retard en la sortida (min)  
Origin: Codi IATA de l'aeroport d'origen  
Dest: Codi IATA de l'aeroport de destí  
Distance: Distància (milles)  
TaxiIn: Rodatge a pista (min)  
TaxiOut: Rodatge a porta (min)  
Cancelled: Si el vol ha sigut o no cancel·lat (1 = si, 0 = no)  
CancellationCode: Codi amb el motiu de la cancel·lació (A = operadora, B = clima, C = NAS, D = seguretat)  
Diverted: Desviat (1 = si, 0 = no)  
CarrierDelay: Retard degut a l'operador (min)  
WeatherDelay: Retard degut al clima (min)  
NASDelay: Retard degut al NAS (min)  
SecurityDelay: Retard degut a Seguretat (min)  
LateAircraftDelay: Retard acumulat de l'avió (min)

Algunes columnes que veiem que podrien ser prescindibles són: CRSDepTime, FlightNum, TailNum, TaxiIn, TaxiOut, CancellationCode


Comencem mirant si hi han cel·les buides

In [5]:
df_flights.isna().sum()

Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Observem que el nombre de cel·les buides es troba en les següents columnes:

ArrTime &nbsp; 7110  
TailNum &nbsp; 5  
ActualElapsedTime &nbsp; 8387  
CRSElapsedTime &nbsp; 198  
AirTime &nbsp; 8387  
ArrDelay &nbsp; 8387  
TaxiIn &nbsp; 7110  
TaxiOut &nbsp; 455  
CarrierDelay &nbsp; &nbsp; 689270  
WeatherDelay &nbsp; &nbsp; 689270  
NASDelay &nbsp; &nbsp; &nbsp; 689270  
SecurityDelay &nbsp; &nbsp; &nbsp; 689270  
LateAircraftDelay &nbsp; 689270

Després de fer una observació general (ajudant-nos també amb Excel), observem que les columnes ActualElapsedTime, AirTime i ArrDelay tenen 8387 buits. Això és degut a que si el vol és cancelat (cancelled) o desviat (diverted), llavors la columna es queda buida.

In [6]:
cancelled_flights = (df_flights['Cancelled'] == 1).sum()
print(cancelled_flights)
diverted_flights = (df_flights['Diverted'] == 1).sum()
print(diverted_flights)
print(cancelled_flights + diverted_flights)

633
7754
8387


Observem també que les columnes CarrierDelay, WeatherDelay, NASDelay, SecurityDelay i LateAircraftDelay tenen 689270 buits. Aquest nombre és degut a que a aquestes columnes els hi manquen dades si la columna ArrDelay és inferior a 15 minuts, si el vol es cancela i si el vol es desvia.

In [7]:
ontime_flights = (df_flights['ArrDelay'] < 15).sum()
print(ontime_flights)
print(cancelled_flights + diverted_flights + ontime_flights)

680883
689270


Per ajudar-nos, a mes a mes d'utilitzar Excel amb el document csv original, ens hem creat un altre amb totes les columnes per les 8387 files en les que ArrDelay es trobava buit. Aixo ho hem fet amb el següent codi que deixem comentat:

In [8]:
'''
ArrDelay_isna = df_flights[df_flights['ArrDelay'].isna()]
ArrDelay_isna.head(10)
ArrDelay_isna.shape
#to_csv
ArrDelay_isna.to_csv('./ArrDelay_isna.csv', index=False)
'''

"\nArrDelay_isna = df_flights[df_flights['ArrDelay'].isna()]\nArrDelay_isna.head(10)\nArrDelay_isna.shape\n#to_csv\nArrDelay_isna.to_csv('./ArrDelay_isna.csv', index=False)\n"

Ens inclinem per no utilitzar el dropna. Omplirem tot el dataframe amb dades. Per aixo utilitzarem Imputation de la manera com ho hem comentat a la següent cel·la.

In [9]:
# Utilitzem averages pels 8387 valors buits del retard a l'arribada, el temps a l'aire i el temps real de vol:
df_flights['ArrDelay'].fillna(df_flights['ArrDelay'].mean(), inplace = True)
df_flights['AirTime'].fillna(df_flights['AirTime'].mean(), inplace = True)
df_flights['ActualElapsedTime'].fillna(df_flights['ActualElapsedTime'].mean(), inplace = True)

# Alla on no tenim el temps de vol programat (CRSElapsedTime), utilitzem el valor del temps real (ActualElapsedTime):
df_flights['CRSElapsedTime'].fillna(df_flights['ActualElapsedTime'], inplace = True)

# Utilitzarem la moda per omplir les columnes TaxiIn i TaxiOut:
df_flights['TaxiIn'].fillna(df_flights['TaxiIn'].mode()[0], inplace = True)
df_flights['TaxiOut'].fillna(df_flights['TaxiOut'].mode()[0], inplace = True)

# Per l'hora d'arribada utilitzem la mediana:
df_flights['ArrTime'].fillna(df_flights['ArrTime'].median(), inplace = True)

# Utilitzarem zeros per omplir els temps de retard dels diferents conceptes. 
# Ho fem degut a que en tots aquests casos els retards estan per sota de 15 minuts per 
# tant considerem aquests retards menystenibles:
df_flights['CarrierDelay'].fillna(0, inplace = True)
df_flights['WeatherDelay'].fillna(0, inplace = True)
df_flights['NASDelay'].fillna(0, inplace = True)
df_flights['SecurityDelay'].fillna(0, inplace = True)
df_flights['LateAircraftDelay'].fillna(0, inplace = True)

# Omplirem amb "Unknown" les matricules dels avions que no quedaren registrades:
df_flights["TailNum"].fillna("Unknown", inplace = True)

print(df_flights.isna().sum())

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64


## Exercici 2
Fes un informe complet del dataset:

- Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
- Troba quantes dades faltants hi ha per columna.
- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
- Fes una taula de les aerolínies amb més endarreriments acumulats.
- Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
- Aporta allò que consideris rellevant.

En un pas farem seleccionarem les columnes que ens interessen i farem el sampling aleatori de 200.000 files:

In [10]:
df_flights_reduced = df_flights[["UniqueCarrier","AirTime","ArrDelay","DepDelay","Origin","Dest", "ActualElapsedTime", "Distance", "Diverted", "Year", "Month", "DayofMonth", "DayOfWeek"]].sample(n=200000)
df_flights_reduced.shape

(200000, 13)

In [11]:
df_flights_reduced.head(10)

UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
1872991            WN     52.0      41.0      43.0    ABQ  TUS   
1678756            AA    193.0       2.0      21.0    DFW  SJC   
6972036            CO     54.0      47.0      51.0    IAH  MFE   
5452445            WN    226.0      15.0       8.0    SEA  MDW   
6980850            CO     42.0      11.0      25.0    IAH  DFW   
3264135            US     53.0       6.0       6.0    LAX  PHX   
813337             UA    144.0      12.0      12.0    DEN  SFO   
1696263            AA     88.0     108.0     127.0    BNA  LGA   
3632593            WN     37.0       6.0      10.0    MCI  TUL   
1680413            AA    138.0     223.0     202.0    ORD  MSY   

         ActualElapsedTime  Distance  Diverted  Year  Month  DayofMonth  \
1872991               63.0       321         0  2008      4          28   
1678756              211.0      1438         0  2008      3          23   
6972036               71.0       316         0  2008     12          30   
5452445              242.0      1733         0  2008     10          23   
6980850               60.0       224         0  2008     12          19   
3264135               86.0       370         0  2008      6           8   
813337               160.0       967         0  2008      2          13   
1696263              106.0       764         0  2008      3           9   
3632593               46.0       223         0  2008      7          11   
1680413              161.0       837         0  2008      3          27   

         DayOfWeek  
1872991          1  
1678756          7  
6972036          2  
5452445          4  
6980850          5  
3264135          7  
813337           3  
1696263          7  
3632593          5  
1680413          4

In [12]:
df_flights_reduced.tail(10)

UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
6159274            UA    130.0      -8.0       7.0    TPA  ORD   
4017658            MQ     43.0       3.0      10.0    DFW  LBB   
801850             OO     32.0      82.0      64.0    SAN  LAX   
1301943            XE     77.0      84.0      75.0    LIT  IAH   
6884204            9E    141.0      12.0      28.0    DTW  OKC   
1492255            DL     84.0      -5.0       8.0    DCA  ATL   
6741643            DL     84.0      47.0      38.0    SRQ  ATL   
1386235            OO     60.0     105.0     104.0    MRY  LAX   
778757             OO    123.0      12.0      15.0    HPN  ORD   
2125586            EV    116.0      19.0       9.0    ATW  ATL   

         ActualElapsedTime  Distance  Diverted  Year  Month  DayofMonth  \
6159274              163.0      1012         0  2008     11          11   
4017658               58.0       282         0  2008      7          19   
801850                66.0       109         0  2008      2          24   
1301943               98.0       374         0  2008      3          17   
6884204              163.0       900         0  2008     12          22   
1492255              101.0       547         0  2008      3          25   
6741643              109.0       445         0  2008     12          17   
1386235               84.0       267         0  2008      3          25   
778757               159.0       738         0  2008      2           9   
2125586              137.0       765         0  2008      4          18   

         DayOfWeek  
6159274          2  
4017658          6  
801850           7  
1301943          1  
6884204          1  
1492255          2  
6741643          3  
1386235          2  
778757           6  
2125586          5

L'anàlisis estadític és el següent:

In [13]:
df_flights_reduced.describe().round(2)

AirTime   ArrDelay   DepDelay  ActualElapsedTime   Distance  \
count  200000.00  200000.00  200000.00          200000.00  200000.00   
mean      108.08      42.19      43.17             133.07     764.32   
std        68.37      56.54      53.20              71.82     573.73   
min         0.00     -67.00       6.00              17.00      11.00   
25%        58.00       9.00      12.00              80.00     337.00   
50%        90.00      25.00      24.00             116.00     606.00   
75%       137.00      56.00      53.00             164.00     997.00   
max       637.00    1417.00    1428.00             667.00    4962.00   

        Diverted      Year      Month  DayofMonth  DayOfWeek  
count  200000.00  200000.0  200000.00   200000.00  200000.00  
mean        0.00    2008.0       6.11       15.75       3.98  
std         0.06       0.0       3.49        8.77       2.00  
min         0.00    2008.0       1.00        1.00       1.00  
25%         0.00    2008.0       3.00        8.00       2.00  
50%         0.00    2008.0       6.00       16.00       4.00  
75%         0.00    2008.0       9.00       23.00       6.00  
max         1.00    2008.0      12.00       31.00       7.00

Veient les estadístiques de retard a la sortida (departure) i a l'arribada (arrival), sembla ser que que una gran part a l'arribada s'arrossega del retard a la sortida. Més endavant farem una correlacio.

Per trobar quantes dades falten per columna executem el següent (hauria d'estar tot net).  

In [14]:
df_flights_reduced.isna().sum()

UniqueCarrier        0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
ActualElapsedTime    0
Distance             0
Diverted             0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
dtype: int64

Crearem 3 columnes: una amb la velocitat mitjana en milles per hora (AvgSpeed), una altra (Late) a on marquem si el vol te un retard de mes de 15 minuts i una altra (Date) amb la data a partir de les columnes de dia, mes i any.

In [15]:
# Velocitat mitjana en milles per hora
df_flights_reduced["AvgSpeed"] = (df_flights_reduced["Distance"] * 60 / df_flights_reduced["ActualElapsedTime"]).round(2)

In [16]:
# Ha tingut el vol un retard de més de 15 min? (1 = Si, 0 = no)
import numpy as np
df_flights_reduced["Late"] = np.where((df_flights_reduced['ArrDelay'] > 15), 1, 0)

In [17]:
# Columna amb data YYYY-M-D
df_flights_reduced['Date'] = pd.to_datetime(dict(year=df_flights_reduced.Year, month=df_flights_reduced.Month, day=df_flights_reduced.DayofMonth))

In [18]:
df_flights_reduced.head(10)

UniqueCarrier  AirTime  ArrDelay  DepDelay Origin Dest  \
1872991            WN     52.0      41.0      43.0    ABQ  TUS   
1678756            AA    193.0       2.0      21.0    DFW  SJC   
6972036            CO     54.0      47.0      51.0    IAH  MFE   
5452445            WN    226.0      15.0       8.0    SEA  MDW   
6980850            CO     42.0      11.0      25.0    IAH  DFW   
3264135            US     53.0       6.0       6.0    LAX  PHX   
813337             UA    144.0      12.0      12.0    DEN  SFO   
1696263            AA     88.0     108.0     127.0    BNA  LGA   
3632593            WN     37.0       6.0      10.0    MCI  TUL   
1680413            AA    138.0     223.0     202.0    ORD  MSY   

         ActualElapsedTime  Distance  Diverted  Year  Month  DayofMonth  \
1872991               63.0       321         0  2008      4          28   
1678756              211.0      1438         0  2008      3          23   
6972036               71.0       316         0  2008     12          30   
5452445              242.0      1733         0  2008     10          23   
6980850               60.0       224         0  2008     12          19   
3264135               86.0       370         0  2008      6           8   
813337               160.0       967         0  2008      2          13   
1696263              106.0       764         0  2008      3           9   
3632593               46.0       223         0  2008      7          11   
1680413              161.0       837         0  2008      3          27   

         DayOfWeek  AvgSpeed  Late       Date  
1872991          1    305.71     1 2008-04-28  
1678756          7    408.91     0 2008-03-23  
6972036          2    267.04     1 2008-12-30  
5452445          4    429.67     0 2008-10-23  
6980850          5    224.00     0 2008-12-19  
3264135          7    258.14     0 2008-06-08  
813337           3    362.62     0 2008-02-13  
1696263          7    432.45     1 2008-03-09  
3632593          5    290.87     0 2008-07-11  
1680413          4    311.93     1 2008-03-27

Fes una taula de les aerolínies amb més endarreriments acumulats.

In [19]:
x = df_flights_reduced[['Late', 'UniqueCarrier']].copy()
late_carriers = x.groupby('UniqueCarrier').aggregate(sum)
late_carriers.sort_values(by='Late', axis=0, ascending=False)

Late
UniqueCarrier       
WN             20391
AA             13456
MQ              9961
UA              9336
OO              9190
DL              7229
XE              7162
US              5996
CO              5921
EV              5793
NW              5557
YV              5225
FL              4756
OH              4098
B6              3976
9E              3591
AS              2479
F9              1538
HA               413
AQ                36

En el següent enllaç de la Federal Aviation Administration trobem els codis corresponent a cada companyia aeria:  
https://aspm.faa.gov/aspmhelp/index/ASQP___Carrier_Codes_And_Names.html

Veiem que la companyia que més endarreriments acumula (amb diferencia) és WN, que és Southwest Airlines i la segona és AA, que correspon a American Airlines.

Els vols més llargs (pel que fa al temps en minuts) serien:

In [20]:
longest_flights = df_flights_reduced[["Origin", "Dest", "AirTime"]]
longest_flights = longest_flights.sort_values(by="AirTime", ascending=False)
longest_flights.head(10)

Origin Dest  AirTime
6971949    EWR  HNL    637.0
576207     EWR  HNL    631.0
570303     EWR  HNL    626.0
1151472    EWR  HNL    622.0
6432868    EWR  HNL    622.0
2354645    EWR  HNL    619.0
3557099    EWR  HNL    618.0
567811     EWR  HNL    616.0
6972756    EWR  HNL    613.0
6988941    EWR  HNL    610.0

Pel que fa als codis dels aeroports, els podem consultar en el següent enllaç:  
https://en.wikipedia.org/wiki/List_of_airports_in_the_United_States  
EL vol més llarg (en temps) és entre Seattle–Tacoma International Airport (SEA) i Honolulu (HNL). També és remarcable el vol entre Newark Liberty International Airport (EWR) i Honolulu (HNL).

Els vols més llargs pel que fa a distancia (en milles) serien els següents, però no ofereix una reposta satisfactoria. Ho resoldrem més tard a l'apartat de les rutes més llargues.

In [21]:
longest_flights_dist = df_flights_reduced[["Origin", "Dest", "Distance"]]
longest_flights_dist = longest_flights_dist.sort_values(by="Distance", ascending=False)
longest_flights_dist.head(10)

Origin Dest  Distance
2355589    HNL  EWR      4962
2354645    EWR  HNL      4962
6971949    EWR  HNL      4962
4805901    EWR  HNL      4962
4820496    EWR  HNL      4962
5361577    EWR  HNL      4962
6432868    EWR  HNL      4962
6972756    EWR  HNL      4962
567811     EWR  HNL      4962
1151472    EWR  HNL      4962

Les rutes més endarrerides (en minuts) serien els següents:

In [22]:
most_delayed_route = df_flights_reduced[["Origin", "Dest", "ArrDelay"]]
most_delayed_route = most_delayed_route.sort_values(by="ArrDelay", ascending=False)
most_delayed_route.head(10)

Origin Dest  ArrDelay
6842014    DSM  ORD    1417.0
1097934    RSW  STL    1312.0
2820874    TPA  DTW    1263.0
5273343    TUS  DFW    1186.0
2837951    BTR  MEM    1177.0
1674045    PSP  DFW    1142.0
6920210    SNA  DFW    1134.0
3378038    LAX  XNA    1129.0
6903387    SEA  DFW    1112.0
3431899    PDX  MSP    1059.0

Observem com l'aeroport de Honolulu (HNL) apareix molt sovint, així com el de Seattle–Tacoma International Airport (SEA). Propablement per raons geogràfiques.

Les rutes més llargues en milles serien:

In [23]:
j = df_flights_reduced[['Distance', 'Origin', 'Dest']].copy()
longest_routes = j.groupby(['Origin', 'Dest'], sort=False)
longest_routes.first().sort_values(by='Distance', axis=0, ascending=False).head(10)

Distance
Origin Dest          
EWR    HNL       4962
HNL    EWR       4962
ATL    HNL       4502
HNL    ATL       4502
       ORD       4243
ORD    HNL       4243
KOA    ORD       4213
ORD    OGG       4184
MSP    HNL       3972
HNL    MSP       3972

El problema de resoldre aquest apartat d'aquesta manera és que ens apareix el mateix vol dues vegades pel fet de que tots dos aeroports són origen i destí :(

Farem també un cop d'ull a la distribucio de retrasos per dia de la setmana i per mes. També fem una correlació entre el retard a la sortida i a l'arribada.

In [24]:
y = df_flights_reduced[['Late', 'DayOfWeek']].copy()
late_per_weekday = y.groupby('DayOfWeek').aggregate(sum)
late_per_weekday.sort_values(by='Late', axis=0, ascending=False)

Late
DayOfWeek       
5          21716
7          18804
4          18775
1          18741
2          17181
3          17122
6          13765

In [25]:
z = df_flights_reduced[['Late', 'Month']].copy()
late_per_month = z.groupby('Month').aggregate(sum)
late_per_month.sort_values(by='Late', axis=0, ascending=False)

Late
Month       
12     14518
6      13795
3      13245
2      13087
1      12288
7      12224
8      10093
4       9749
5       9424
11      6510
10      5694
9       5477

In [26]:
def corr_coef(arr1, arr2):
    return(np.corrcoef(arr1, arr2))

print(corr_coef(df_flights_reduced["ArrDelay"], df_flights_reduced["DepDelay"]))

[[1.         0.94868765]
 [0.94868765 1.        ]]


- Podem veure que hi ha una correlacio lineal positiva bastant forta entre el retard a la sortida i a l'arribada.
- També sembla ser que els divendres és el dia que s'acumulen més retards i el segon és dilluns, podria ser que tingues alguna cosa a veure amb el cap de setmana.
- Pel que fa a la ditribució per mesos no hi ha una tendencia molt clara, tot i que s'aprecia una etapa entre setembre i novembre amb un mínim de retrasos, probablement per haver menos vols en aquesta etapa de l'any, però s'hauria de comprobar.

## Exercici 3
Exporta el dataset net i amb les noves columnes a Excel.

In [27]:
# Conservarem l'index (per treure'l, index=False)
df_flights_reduced.to_excel("./DelayedFlightsWorked.xlsx", index = True)